kvserver: add mmaStoreRebalancer

wenyihu6 opened the pull request:
**kvserver: add MMARangeLoad**

Previously, we lacked a utility method for mma to inspect range load and convert
it to a mmaprototype.RangeLoad. This commit adds MMARangeLoad(), which builds a
mmaprototype.RangeLoad from a replica’s LoadStats. It is currently unused;
future commits will use it to populate the range message field for MMA input.

Epic: none
Release note: none

---

**kvserver: add mmaRangeMessageNeeded**

Previously, there was no struct under Replica to track state changes and signal
mma when an up-to-date range message with full information was needed. This
commit introduces such a struct - mmaRangeMessageNeeded. It is marked as needed
when any of the following occur: replica initialization, span config changes,
leaseholder changes, or range descriptor updates.

Epic: none
Release note: none

---

**kvserver: add more methods to mmaRangeMessageNeeded**

Previously, mmaRangeMessageNeeded was introduced to track when a replica state
change required sending updated information. This commit adds methods to
mmaRangeMessageNeeded to help mma decide whether a new range message with full
information is needed. Since processing full updates can be expensive, mma sends
them only when necessary. In addition to tracking state changes, the struct now
also tracks the content of the last full info range message sent to determine if
full information is necessary.

Epic: none
Release note: none

---

**kvserver: add store.MakeStoreLeaseholderMsg**

Previously, mma lacked a method to retrieve range messages for all leaseholder
replicas on a store. This commit adds such a method to Store, which iterates
over all replicas and constructs range message for leaseholder replicas. Note
that it is currently left unused. mma will use this as input for computing
rebalancing decisions in future commits.

Epic: none
Release note: none

---

**kvserver: add mmaStoreRebalancer**

This commit introduces mma_store_rebalancer.go, which defines the main struct
mmaStoreRebalancer and its basic methods for computing and applying rebalancing
changes. It is currently unplumbed; future commits will wire it into each store
and start a background goroutine to perform rebalancing periodically.

Epic: none
Release note: none

---

**kvserver: plumb mmaStoreRebalancer**

Previously, mmaStoreRebalancer was unplumbed. This commit initializes it per
Store. It also constructs and plumbs a shared MMAAllocator per node, which is
shared by all stores' mmaStoreRebalancer on the same node.

Epic: none
Release note: none

---

**kvserver: use RegisterCallback for mma**

Previously, we extended RegisterCallback to include origTimestampNanos in the
callback arguments. This commit puts it to use for MMA by registering a callback
in NewServer that notifies the MMA allocator of stores and updated store load
messages. This callback is triggered when the store descriptor is gossiped via
StoreGossip, either periodically or due to significant store capacity changes.

Epic: none
Release note: none

wenyihu6 committed:
kvserver: add MMARangeLoad

Previously, we lacked a utility method for mma to inspect range load and convert
it to a mmaprototype.RangeLoad. This commit adds MMARangeLoad(), which builds a
mmaprototype.RangeLoad from a replica’s LoadStats. It is currently unused;
future commits will use it to populate the range message field for MMA input.

Epic: none
Release note: none

wenyihu6 committed:
kvserver: add mmaRangeMessageNeeded

Previously, there was no struct under Replica to track state changes and signal
mma when an up-to-date range message with full information was needed. This
commit introduces such a struct - mmaRangeMessageNeeded. It is marked as needed
when any of the following occur: replica initialization, span config changes,
leaseholder changes, or range descriptor updates.

Epic: none
Release note: none

wenyihu6 committed:
kvserver: add store.MakeStoreLeaseholderMsg

Previously, mma lacked a method to retrieve range messages for all leaseholder
replicas on a store. This commit adds such a method to Store, which iterates
over all replicas and constructs range message for leaseholder replicas. Note
that it is currently left unused. mma will use this as input for computing
rebalancing decisions in future commits.

Epic: none
Release note: none

wenyihu6 committed:
kvserver: add mmaStoreRebalancer

This commit introduces mma_store_rebalancer.go, which defines the main struct
mmaStoreRebalancer and its basic methods for computing and applying rebalancing
changes. It is currently unplumbed; future commits will wire it into each store
and start a background goroutine to perform rebalancing periodically.

Epic: none
Release note: none

wenyihu6 committed:
kvserver: plumb mmaStoreRebalancer

Previously, mmaStoreRebalancer was unplumbed. This commit initializes it per
Store. It also constructs and plumbs a shared MMAAllocator per node, which is
shared by all stores' mmaStoreRebalancer on the same node.

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/150780)
<!-- Reviewable:end -->

wenyihu6 committed:
kvserver: add more methods to mmaRangeMessageNeeded

Previously, mmaRangeMessageNeeded was introduced to track when a replica state
change required sending updated information. This commit adds methods to
mmaRangeMessageNeeded to help mma decide whether a new range message with full
information is needed. Since processing full updates can be expensive, mma sends
them only when necessary. In addition to tracking state changes, the struct now
also tracks the content of the last full info range message sent to determine if
full information is necessary.

Epic: none
Release note: none

wenyihu6 committed:
kvserver: use RegisterCallback for mma

Previously, we extended RegisterCallback to include origTimestampNanos in the
callback arguments. This commit puts it to use for MMA by registering a callback
in NewServer that notifies the MMA allocator of stores and updated store load
messages. This callback is triggered when the store descriptor is gossiped via
StoreGossip, either periodically or due to significant store capacity changes.

Epic: none
Release note: none

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150780)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/kv/kvserver/replica.go` line 2987 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/150780#-OWE6UDWDSdhVGsAs7wK:-OWE6UDWDSdhVGsAs7wL:btfqgcb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40053ad7a38d875a7f04a02798ba93cd4b96fc3/pkg/kv/kvserver/replica.go#L2987)):*
> ```Go
> // - when significant range load changes since the last RangeMessage constructed
> type mmaRangeMessageNeeded struct {
> 	// TODO(wenyihu6): highlight the mutex diff during review
> ```

The [prototype](https://github.com/tbg/cockroach/blob/06bb789bbe567373a534ab8fe558bc5c798886fe/pkg/kv/kvserver/replica.go#L3106) has `mmaRangeMessageNeeded` protected by `r.mu`. It seems unnecessary to me since only `stateChangeTriggered` is accessed concurrently. And it should be fine for MMA to observe a slightly stale replica state. 
___
*[`pkg/kv/kvserver/replica.go` line 3069 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/150780#-OWE7Oz4AKGbdMEpJBSr:-OWE7Oz4AKGbdMEpJBSs:br5m9pg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40053ad7a38d875a7f04a02798ba93cd4b96fc3/pkg/kv/kvserver/replica.go#L3069)):*
> ```Go
> 		// here, since it is stronger than the lack of a send-queue. And we needed
> 		// the send-queue state for something in the allocator.
> 		// TODO(wenyihu6 during review): highlight diff w prototype here
> ```

This also differs slightly from the prototype https://github.com/tbg/cockroach/blob/06bb789bbe567373a534ab8fe558bc5c798886fe/pkg/kv/kvserver/replica.go#L3024, which didn’t check the map’s length or whether the replica ID exists.
___
*[`pkg/kv/kvserver/replica.go` line 3150 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/150780#-OWE8OhOCPuVo6JpEOxP:-OWE8OhOCPuVo6JpEOxQ:b-pv3fnr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40053ad7a38d875a7f04a02798ba93cd4b96fc3/pkg/kv/kvserver/replica.go#L3150)):*
> ```Go
> 	// be lost before calling into MMA anyway. Highlight the diff with prototype
> 	// here during review.
> 	isLeaseholder := status.IsValid() && status.Lease.OwnedBy(r.store.StoreID())
> ```

This also differs slightly from the prototype which directly accesses `r.shMu.state.Lease` https://github.com/tbg/cockroach/blob/06bb789bbe567373a534ab8fe558bc5c798886fe/pkg/kv/kvserver/replica.go#L3103. I previously asked Sumeer about it: https://cockroachlabs.slack.com/archives/C048HDZJSAY/p1750709824092029?thread_ts=1750686091.586809&cid=C048HDZJSAY. It was done that way since it was cheaper. I switched to the more expensive but more standard approach here. We can revert to the cheaper version later if we're confident it’s safe.
___
*[`pkg/kv/kvserver/replica.go` line 3231 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/150780#-OWE96DW2e21cGD2W9ZI:-OWE96DW2e21cGD2W9ZJ:b-guj69a) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40053ad7a38d875a7f04a02798ba93cd4b96fc3/pkg/kv/kvserver/replica.go#L3231)):*
> ```Go
> 	// lock while clearing mmaRangeMessageNeeded and setting wasLeaseholder which
> 	// seems unnecessary since we are the only one accessing them and stale view
> 	// should be fine for mma.
> ```

Prototype held the r.mu lock when clearing state and changing mmaRangeMessageNeeded https://github.com/tbg/cockroach/blob/06bb789bbe567373a534ab8fe558bc5c798886fe/pkg/kv/kvserver/replica.go#L3105-L3108 which seemed unnecessary to me. 
___
*[`pkg/kv/kvserver/replica.go` line 3262 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/150780#-OWE9PLC944qMnrnEVZv:-OWE9PLC944qMnrnEVZw:b39k10p) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40053ad7a38d875a7f04a02798ba93cd4b96fc3/pkg/kv/kvserver/replica.go#L3262)):*
> ```Go
> 	// TODO(wenyihu6): check if we can use r.raftSparseStatusRLocked() here
> 	// (cheaper).
> 	// TODO(wenyihu6): highlight the diff with prototype here. Prototype hold r.mu
> ```

Same here with the lock situation. 


<!-- Sent from Reviewable.io -->
